substrate-node/pallets/pallet-tfgrid/src/lib.rs

```diff
@@ -2327,6 +2368,70 @@ impl<T: Config> Pallet<T> {
         let parsed_serial_number = <T as Config>::SerialNumber::try_from(serial_number)?;
         Ok(parsed_serial_number)
     }
+
+    fn _change_power_state(
+        account_id: T::AccountId,
+        power_state: Power,
+    ) -> DispatchResultWithPostInfo {
+        let twin_id = TwinIdByAccountID::<T>::get(&account_id).ok_or(Error::<T>::TwinNotExists)?;
+        ensure!(
+            NodeIdByTwinID::<T>::contains_key(twin_id),
+            Error::<T>::NodeNotExists
+        );
+        let node_id = NodeIdByTwinID::<T>::get(twin_id);
+        let node = Nodes::<T>::get(node_id).ok_or(Error::<T>::NodeNotExists)?;
+
+        let power_state = match power_state {
+            Power::Up => PowerState::Up,
+            Power::Down => PowerState::Down(system::Pallet::<T>::block_number()),
+        };
+
+        let mut node_power = NodePower::<T>::get(node_id);
+
+        // if the power state is not correct => change it and emit event
+        if node_power.state != power_state {
+            node_power.state = power_state.clone();
+
+            NodePower::<T>::insert(node_id, node_power);
+            Self::deposit_event(Event::PowerStateChanged {
+                farm_id: node.farm_id,
+                node_id,
+                power_state,
+            });
+        }
+
+        Ok(Pays::No.into())
+    }
+
+    fn _change_power_target(
+        account_id: T::AccountId,
+        node_id: u32,
+        power_target: Power,
+    ) -> DispatchResultWithPostInfo {
+        let twin_id = TwinIdByAccountID::<T>::get(&account_id).ok_or(Error::<T>::TwinNotExists)?;
+        let node = Nodes::<T>::get(node_id).ok_or(Error::<T>::NodeNotExists)?;
+        let farm = Farms::<T>::get(node.farm_id).ok_or(Error::<T>::FarmNotExists)?;
+        ensure!(
+            twin_id == farm.twin_id,
+            Error::<T>::UnauthorizedToChangePowerTarget
+        );
+
```

Comment:
Shouldn't we check if node has active contract/deployment on it in case of state is `Up` and farmer ask to shut it `Down`?

Reply:
Indeed. The question is what do we do with those contracts then? Cancel them or throw an error... I like the idea of throwing an error so that we don't cancel them.

Reply:
The farmerbot should be doing this check, it doesn't really matter to check it here in my opinion. 

The way I see it is that a farmer always has to opportunity to physically shut down his node thus having the same possibility on chain without consequences is also an option. Also, we don't want to drag pallet-smart-contract as a dependency in this tfgrid pallet.

Reply:
That's also true.